src/librustc_target/abi/mod.rs

```diff
@@ -871,8 +871,26 @@ impl Niche {
 
 #[derive(PartialEq, Eq, Hash, Debug, HashStable_Generic)]
 pub struct LayoutDetails {
-    pub variants: Variants,
+    /// Says where the fields are located.
+    /// Primitives and fieldless enums appear as unions without fields.
     pub fields: FieldPlacement,
+
+    /// Encodes information about multi-variant layouts.
+    /// Even with `Multiple` variants, a layout can still have fields! Those are then
+    /// shared between all variants. One of them will be the discriminant,
+    /// but e.g. generators can have more.
+    ///
+    /// A layout-guided recursive descent must first look at all the fields,
+    /// and only then check if this is a multi-variant layout and if so, proceed
+    /// with the active variant.
```

Comment:
I'm not sure what to do with this recursive descent documentation, I don't mind it too much, but it feels out of place.

I'm trying to think of something more general, such as talking about "working with values of this layout" or "accessing variant fields" requiring inspecting the discriminant to find the active variant.

Reply:
I did some slight editing here, please have a look.

Reply:
LOoks really great, thanks!